packages/bundlers/experimental/src/ExperimentalBundler.js

```diff
@@ -895,39 +892,57 @@ function createIdealGraph(
 
     // Finally, filter out bundleRoots (bundles) from this assets
     // reachable if they are subgraphs, and reuse that subgraph bundle
-    // by drawing an edge. Essentially, if two bundles within an assets
+    // by drawing an edge. Essentially, if two bundles within an asset's
     // reachable array, have an ancestor-subgraph relationship, draw that edge.
     // This allows for us to reuse a bundle instead of making a shared bundle if
     // a bundle represents the exact set of assets a set of bundles would share
 
     // if a bundle b is a subgraph of another bundle f, reuse it, drawing an edge between the two
-    reachable = reachable.filter(b => {
-      if (b.env.isIsolated()) {
-        return true;
-      }
-      let toKeep = true;
-      if (bundles.has(asset.id)) {
-        toKeep = false;
-        bundleGraph.addEdge(
-          nullthrows(bundles.get(b.id)),
-          nullthrows(bundles.get(asset.id)),
-        );
+    let canReuse: Set<BundleRoot> = new Set();
+    for (let candidateSourceBundleRoot of reachable) {
+      let candidateSourceBundleId = nullthrows(
+        bundles.get(candidateSourceBundleRoot.id),
+      );
+      if (candidateSourceBundleRoot.env.isIsolated()) {
+        continue;
       }
-      for (let f of reachable) {
-        if (b === f) continue;
-        let fReachable = getReachableBundleRoots(f, reachableRoots).filter(
-          b => !ancestorAssets.get(b)?.has(f),
-        );
-        if (fReachable.indexOf(b) > -1) {
-          toKeep = false;
-          bundleGraph.addEdge(
-            nullthrows(bundles.get(b.id)),
-            nullthrows(bundles.get(f.id)),
-          );
+      let reuseableBundleId = bundles.get(asset.id);
+      if (reuseableBundleId != null) {
+        canReuse.add(candidateSourceBundleRoot);
+        bundleGraph.addEdge(candidateSourceBundleId, reuseableBundleId);
+
+        let reusableBundle = bundleGraph.getNode(reuseableBundleId);
+        invariant(reusableBundle !== 'root' && reusableBundle != null);
+        reusableBundle.sourceBundles.add(candidateSourceBundleId);
+      } else {
+        // Asset is not a bundleRoot, but if its parent bundle can be
+        // reused as a subgraph of another bundle in its reachable, reuse it
+        for (let otherReuseCandidate of reachable) {
```

Comment:
So the algorithm here went from O(n) to O(n^2)? Did we see any real-world perf regressions?

Reply:
nevermind, this was there before